CRM/Member/Form/Membership.php

```diff
@@ -726,110 +726,51 @@ public static function formRule($params, $files, $self) {
-
-            if (!empty($params['is_override']) && !CRM_Member_StatusOverrideTypes::isPermanent($params['is_override'])) {
-              $errors['is_override'] = ts('Because you set an End Date for a lifetime membership, This must be set to "Override Permanently"');
-            }
-          }
-          else {
-            if (!$startDate) {
-              $errors['start_date'] = ts('Start date must be set if end date is set.');
-            }
-            if ($endDate < $startDate) {
-              $errors['end_date'] = ts('End date must be the same or later than start date.');
-            }
-          }
-        }
-
-        // Default values for start and end dates if not supplied on the form.
-        $defaultDates = CRM_Member_BAO_MembershipType::getDatesForMembershipType($memType,
-          $joinDate,
-          $startDate,
-          $endDate
+        $defaultDates = \CRM_Member_BAO_MembershipType::getDatesForMembershipType($memType,
+          $params['join_date'] ?? NULL,
+          $params['start_date'] ?? NULL,
+          $params['end_date'] ?? NULL
+        );
+        $calcStatus = \CRM_Member_BAO_MembershipStatus::getMembershipStatusByDate(
+          $params['start_date'] ?? CRM_Utils_Array::value('start_date', $defaultDates),
+          $params['end_date'] ?? CRM_Utils_Array::value('end_date', $defaultDates),
+          $params['join_date'] ?? CRM_Utils_Array::value('join_date', $defaultDates),
+          'now',
+          TRUE,
+          $memType
         );
 
-        if (!$startDate) {
-          $startDate = CRM_Utils_Array::value('start_date',
-            $defaultDates
-          );
-        }
-        if (!$endDate) {
-          $endDate = CRM_Utils_Array::value('end_date',
-            $defaultDates
-          );
-        }
-
-        //CRM-3724, check for availability of valid membership status.
-        if ((empty($params['is_override']) || CRM_Member_StatusOverrideTypes::isNo($params['is_override'])) && !isset($errors['_qf_default'])) {
-          $calcStatus = CRM_Member_BAO_MembershipStatus::getMembershipStatusByDate($startDate,
-            $endDate,
-            $joinDate,
-            'now',
-            TRUE,
-            $memType,
-            $params
-          );
-          if (empty($calcStatus)) {
-            $url = CRM_Utils_System::url('civicrm/admin/member/membershipStatus', 'reset=1&action=browse');
-            $errors['_qf_default'] = ts('There is no valid Membership Status available for selected membership dates.');
-            $status = ts('Oops, it looks like there is no valid membership status available for the given membership dates. You can <a href="%1">Configure Membership Status Rules</a>.', [1 => $url]);
-            if (!$self->_mode) {
-              $status .= ' ' . ts('OR You can sign up by setting Status Override? to something other than "NO".');
+        $validationErrors = civicrm_api4('Membership', 'validate', [
+          'values' => [
+            array_merge($params, [
```

Comment:
I would avoid doing the `array_merge()` here & only pass out the specific parameters you are currently adding validate for. That reduces the risk someone builds an implementation that deals with a value that 'just happens to be in params' - e.g an apiv3 style custom field or something specific to the form without clarifying the use case & adding a unit test.   As you know people almost never extend CiviCRM in weird and wonderful ways that depends on some historical piece of coding setting some odd value that happened to be there the month they wrote their customisation and then expect it to always work with no testing .... but it's a possibility....